### PR TITLE
fix!: load asset images for tests

### LINKED ISSUE #16: Assets image displayed for CI snapshot but not for platform
### Is there an existing issue for this?

- [X] I have searched the existing issues.

### Version

0.2.1

### Description

When I'm generating golden files using an asset image it is not rendered while it is rendered correctly for its ci equivalent.

### Steps to reproduce

1. Create a `TestAssetBundle` to load your assets

```dart
import 'dart:convert';

import 'package:flutter/services.dart';

class TestAssetBundle extends CachingAssetBundle {
  @override
  Future<String> loadString(String key, {bool cache = true}) async {
    final data = await load(key);
    return utf8.decode(data.buffer.asUint8List());
  }

  @override
  Future<ByteData> load(String key) => rootBundle.load(key);
}
```

2. Write a `goldenTest` with an `Image.asset` widget (wrapped inside a `DefaultAssetBundle`)

```dart
import 'package:alchemist/alchemist.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import '../../utils/test_asset_bundle.dart';

void main() {
  setUpAll(TestWidgetsFlutterBinding.ensureInitialized);

  group('Image.asset: Golden Tests', () {
    goldenTest(
      'renders correctly',
      fileName: 'image_asset',
      widget: GoldenTestGroup(
        children: [
          GoldenTestScenario(
            name: 'XYZ Monster.jpg asset',
            child: DefaultAssetBundle(
              bundle: TestAssetBundle(),
              child: Image.asset('assets/type/XYZ Monster.jpg'),
            ),
          ),
        ],
      ),
    );
  });
}
```

3. Run the command `flutter test --update-goldens`

### Expected behavior

The golden file generated should display the asset image.

## Asset Image

XYZ Monster

## CI Golden

image_asset

## Platform Golden

image_asset

### Additional context and comments

* `flutter doctor -v`:

```bash
[√] Flutter (Channel stable, 2.10.2, on Microsoft Windows [version 10.0.22000.527], locale fr-FR)
    • Flutter version 2.10.2 at C:\src\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 097d3313d8 (5 days ago), 2022-02-18 19:33:08 -0600
    • Engine revision a83ed0e5e3
    • Dart version 2.16.1
    • DevTools version 2.9.2

[√] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at C:\Users\Guillaume\AppData\Local\Android\sdk
    • Platform android-30, build-tools 30.0.3
    • Java binary at: D:\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[!] Visual Studio - develop for Windows (Visual Studio Build Tools 2017 15.9.35)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools
    • Visual Studio Build Tools 2017 version 15.9.28307.1500
    • Windows 10 SDK version 10.0.17763.0
    X Visual Studio 2019 or later is required.
      Download at https://visualstudio.microsoft.com/downloads/.
      Please install the "Desktop development with C++" workload, including all of its default components

[√] Android Studio (version 4.1)
    • Android Studio at D:\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [version 10.0.22000.527]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 98.0.4758.102
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 97.0.1072.76

[√] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.
```

### PR DESCRIPTION
**Pull request by Kirpal**
## Description

Closes #16 .

I also made a (breaking) refactor to use a `builder` instead of a `widget`, which should prevent state from being unintentionally maintained between the ci and platform test runs. Also, what do we think about including `precacheImages` in this package?

## Type of Change

- [ ] ✨ New feature (non-breaking change which adds functionality)
- [x] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [x] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: lib/src/golden_test.dart [hunk 1]
```diff
@@ -70,10 +70,10 @@ Future<void> loadFonts() async {
 /// golden image under the `goldens` directory. This name should be unique, and
 /// may not contain an extension (such as `.png`).
 ///
-/// The provided [widget] describes the scenarios and layout of the widgets that
-/// are included in the test. A child must be provided. Alchemist provides two
-/// widgets to make creating a golden test scenario. See [GoldenTestGroup] and
-/// [GoldenTestScenario] for more details.
+/// The provided [builder] describes the scenarios and layout of the widgets
+/// that are included in the test. A child must be provided. Alchemist provides
+/// two widgets to make creating a golden test scenario. See [GoldenTestGroup]
```
Comment (btrautmann) [inline lib/src/golden_test.dart:75]: I don't think this sentence is complete now that I read it 👀

File: test/src/pumps_test.dart [hunk 2]
```diff
@@ -52,5 +84,134 @@ void main() {
...
+                  Image(
+                    image: memoryImage,
+                  ),
+                ],
+              ),
+            ),
+          );
+          await precacheImages(tester);
+
+          expect(
```
Comment (marcossevilla) [inline test/src/pumps_test.dart:116]: can we use `expectLater` here?

File: test/src/pumps_test.dart [hunk 3]
```diff
@@ -52,5 +84,134 @@ void main() {
...
+                    image: memoryImage,
+                    placeholder: MemoryImage(redPixelImage),
+                  ),
+                ],
+              ),
+            ),
+          );
+          await precacheImages(tester);
+
+          expect(
```
Comment (marcossevilla) [inline test/src/pumps_test.dart:151]: same `expectLater` question

File: test/src/pumps_test.dart [hunk 4]
```diff
@@ -52,5 +84,134 @@ void main() {
...
+                      ),
+                    ),
+                  ),
+                ],
+              ),
+            ),
+          );
+          await precacheImages(tester);
+
+          expect(
```
Comment (marcossevilla) [inline test/src/pumps_test.dart:201]: same `expectLater` question

File: test/src/pumps_test.dart [hunk 5]
```diff
@@ -1,9 +1,41 @@
...
+Future<bool> _isCached(ImageProvider<Object> image, Finder findImage) async {
+  final cacheStatus = await image.obtainCacheStatus(
+    configuration: createLocalImageConfiguration(
+      findImage.evaluate().first,
+    ),
+  );
+  return cacheStatus?.keepAlive ?? false;
+}
+
+extension on CommonFinders {
```
Comment (marcossevilla) [inline test/src/pumps_test.dart:20]: thoughts on moving this to a `helpers/finders.dart` file?

**Review by marcossevilla — COMMENTED** [review]
this is looking great! 🔥 had a couple of nit comments

File: test/src/pumps_test.dart (hunk 5 above)
Comment (Kirpal) [reply test/src/pumps_test.dart:20]: I don't think we'll use them outside this file, so I'd rather keep them here.

**Review by marcossevilla — APPROVED** [review]
lgtm 💯

**Pull request merged**